docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -0,0 +1,1946 @@
+the new package is uploaded to a participant with the old version, the
+participant ensures that it can automatically upgrade every change
+between the old and new version before adding it to the package set.
+
+In order to be able to automatically upgrade a contract or datatype, SCU
+restricts the kinds of changes that a new package version can introduce
+over its prior version.
+
+For example, the simplest kind of data transformation that SCU supports
+is adding a field to a template.
+
+Given the following first version of a template:
+
+.. code:: daml
+
+  template IOU
+    with
+      issuer: Party
+      owner: Party
+      value: Int
+    where
+      signatory issuer
+      observer owner
+
+We can add a new field for currency:
+
+.. code:: daml
+
+  template IOU
+    with
+      issuer: Party
+      owner: Party
+      value: Int
+      -- New field:
+      currency: Optional String
+    where
+      signatory issuer
+      observer owner
+
+With SCU, any new template fields must be optional - templates from the
+old version are automatically upgraded to new versions by setting the
+new field to ``None``. This optional field requirement extends to all
+records in your package. Conversely, newer contracts with this field set
+to ``None`` can be automatically downgraded to previous versions of the
+template in workflows that have not yet been updated.
+
+Automatic Data Upgrades
+~~~~~~~~~~~~~~~~~~~~~~~
+
+When extending data in a daml model, SCU requires the old model be
+representable in the new model. I.e., for extending a record, we can
+only add nullable (``Optional``) fields, so that old data can be represented
+by setting these fields to ``None``. This mentality is continued for
+Variants and Enums, which only allow adding constructors, with some
+other restrictions covered in `Continuing to Write Your Upgrades <#continuing-to-write-your-upgrades>`__. This
+approach is inspired by
+`Protobuf <https://protobuf.dev/programming-guides/proto3/#updating>`__
+and Typescript's ability to ignore `excess
+fields <https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks>`__
+via ``as``.
+
+Following is a list of places where automatic data upgrades occur:
+
+**Submissions to the ledger-api**
+
+When you submit a command, and do not explicitly specify a package-id to
+use, canton will automatically upgrade the payloads you give to the most
+recent version of the package that is uploaded on the participant. It
+will also use the most recent implementation of any choices you exercise
+directly through the ledger-api. This behavior can be influenced by
+package preference, which will be covered later.
+
+**Fetches/Creates**
+
+Within choices, the package-ids of the payloads we expect for fetches
+and creates are given by the version of the package we depend on. SCU
+will upgrade (or downgrade) the payloads given/stored in the ledger to
+match these versions when used in a choice body.
```

Comment:
Technical question: can there be a downgrade happening on a `create`?

Reply:
Yes, if the package-preference on the participant/submission chooses the previous version.
This would also happen if V2 was unvetted.

Reply:
> This would also happen if V2 was unvetted.

**Note:** Vetting does not affect the participant package-preference in 2.9 and there's no plan support this at all for 2.x. In 2.9, the general recommendation should be to not rely on the vetting state but, where needed, explicitly select the desired package-id to be used

Reply:
Oh! Okay is this noted in the docs already?

Reply:
> Note: Vetting does not affect the participant package-preference in 2.9 and there's no plan support this at all for 2.x. In 2.9, the general recommendation should be to not rely on the vetting state but, where needed, explicitly select the desired package-id to be used

@basvangijzel-DA @tudor-da : this requires adapting the [upgrading guidelines here](https://docs.google.com/document/d/10ttHyhufCI7hGEsd0CA2VCjsXPhAg7Hg-fVeMQOKS4U/edit?disco=AAABQ23TQm8), let's figure it out together there.

Reply:
> Yes, if the package-preference on the participant/submission chooses the previous version.

This could happen only for ledger API command.

Reply:
@tudor-da can you please confirm if this needs to be written in limitations?

Reply:
I've added a note for this in the limitations section. I'm going to leave this comment unresolved so Bas see's the note about the upgrading guidelines.

Reply:
After discussion, we're removing this section as unvetting is not recommended in 2.x, nor is uploading without vetting.
Given the state of vetting in 2.x, any recommendation we give will be met with more issues.

Reply:
Agreed. Simon already wrote up his notes here: https://docs.google.com/document/d/10ttHyhufCI7hGEsd0CA2VCjsXPhAg7Hg-fVeMQOKS4U/edit#bookmark=id.h6kczypfi7qv